crates/kitsune_p2p/types/src/tx2/tx2_utils/logic_chan.rs

```diff
@@ -0,0 +1,240 @@
+use crate::tx2::tx2_utils::Share;
+use crate::*;
+use futures::future::{BoxFuture, FutureExt};
+use futures::stream::futures_unordered::FuturesUnordered;
+use futures::stream::Stream;
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+enum LType<E: 'static + Send> {
+    Event(E),
+    Logic(OwnedSemaphorePermit, BoxFuture<'static, ()>),
+}
+type LTypeSend<E> = tokio::sync::mpsc::Sender<LType<E>>;
+type LTypeRecv<E> = tokio::sync::mpsc::Receiver<LType<E>>;
+
+struct LogicChanInner<E: 'static + Send> {
+    send: LTypeSend<E>,
+    logic_limit: Arc<Semaphore>,
+}
+
+/// Handle to a logic_chan instance.
+/// A clone of a LogicChanHandle is `Eq` to its origin.
+/// A clone of a LogicChanHandle will `Hash` the same as its origin.
+pub struct LogicChanHandle<E: 'static + Send>(Share<LogicChanInner<E>>);
+
+impl<E: 'static + Send> Clone for LogicChanHandle<E> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<E: 'static + Send> PartialEq for LogicChanHandle<E> {
+    fn eq(&self, oth: &Self) -> bool {
+        self.0.eq(&oth.0)
+    }
+}
+
+impl<E: 'static + Send> Eq for LogicChanHandle<E> {}
+
+impl<E: 'static + Send> std::hash::Hash for LogicChanHandle<E> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+impl<E: 'static + Send> LogicChanHandle<E> {
+    /// Cause the logic_chan to emit an event.
+    pub fn emit(
+        &self,
+        e: E,
+    ) -> impl std::future::Future<Output = KitsuneResult<()>> + 'static + Send {
+        let e = LType::Event(e);
+        let send = self.0.share_mut(|i, _| Ok(i.send.clone()));
```

Comment:
In our benchmarks we saw that cloning channels was one of the slowest operations. Is this nessacary? Maybe we could hold the channel in the handle?

Reply:
Especially with tokio 1 where send is now `&self` instead of `&mut self` so we can wrap it in an Arc, and clone the Arc instead.

Reply:
Hmm, looks like the [tokio 1 sender internally is Arc-ed](https://docs.rs/tokio/1.3.0/src/tokio/sync/mpsc/chan.rs.html#16)... so maybe it's fast to just clone them now?